src/ray/raylet/task_dependency_manager.h

```diff
@@ -16,71 +16,128 @@ class ReconstructionPolicy;
+  /// previously had all of their dependencies fulfilled, but are now missing
+  /// this object dependency.
+  std::vector<TaskID> HandleObjectMissing(const ray::ObjectID &object_id);
 
  private:
-  /// Check whether the given list of objects are ready.
-  bool argumentsReady(const std::vector<ObjectID> arguments) const;
-  /// Handle an object added to the object store.
-  void handleObjectReady(const ray::ObjectID &object_id);
-  /// A reference to the object manager so that we can issue Pull requests of
-  /// missing objects.
-  ObjectManager &object_manager_;
-  /// A mapping from task ID of each subscribed task to its list of
+  using ObjectDependencyMap = std::unordered_map<ray::ObjectID, std::vector<ray::TaskID>>;
+
+  /// A struct to represent the object dependencies of a task.
+  struct TaskDependencies {
+    /// The objects that the task is dependent on. These must be local before
+    /// the task is ready to execute.
+    std::unordered_set<ObjectID> object_dependencies;
+    /// The number of object arguments that are not available locally. This
+    /// must be zero before the task is ready to execute.
+    int64_t num_missing_dependencies;
+  };
+
+  /// Check whether the given object needs to be made available through object
+  /// transfer or reconstruction. These are objects for which: (1) there is a
+  /// subscribed task dependent on it, (2) the object is not local, and (3) the
+  /// task that creates the object is not pending execution locally.
+  bool CheckObjectRequired(const ObjectID &object_id) const;
+  /// If the given object is required, then request that the object be made
+  /// available through object transfer or reconstruction.
+  void HandleRemoteDependencyRequired(const ObjectID &object_id);
+  /// If the given object is no longer required, then cancel any in-progress
+  /// operations to make the object available through object transfer or
+  /// reconstruction.
+  void HandleRemoteDependencyCanceled(const ObjectID &object_id);
+
+  ObjectManagerInterface &object_manager_;
+  /// A mapping from task ID of each subscribed task to its list of object
   /// dependencies.
-  std::unordered_map<ray::TaskID, std::vector<ray::ObjectID>> task_dependencies_;
-  // A mapping from object ID of each object that is not locally available to
-  // the list of subscribed tasks that are dependent on it.
-  std::unordered_map<ray::ObjectID, std::vector<ray::TaskID>> remote_object_dependencies_;
-  // The set of locally available objects.
+  std::unordered_map<ray::TaskID, TaskDependencies> task_dependencies_;
+  /// All tasks whose outputs are required by a subscribed task. This is a
+  /// mapping from task ID to information about the objects that the task
+  /// creates, either by return value or by `ray.put`. For each object, we
+  /// store the IDs of the subscribed tasks that are dependent on the object.
+  std::unordered_map<ray::TaskID, ObjectDependencyMap> required_tasks_;
```

Comment:
Yeah, I'm not super happy with this data structure either...

I think you understand it correctly. The reason for having the nested map is, as you said, so that we can iterate over the `ObjectIDs` produced by a given task. It's really just an optimization for the code in `TaskPending` and `TaskCanceled`, so that we can request/cancel objects produced by a task if that task becomes pending or canceled locally. Otherwise, it would've been okay to just have a single `ObjectDependencyMap`.

The alternative is to have a single `ObjectDependencyMap` and then a separate map from task ID to object IDs produced by the task. I didn't really like having to update two maps, though. Undecided on which one ends up being cleaner in the end; do you have a preference?

Reply:
Not sure, why don't we try it this way and revisit it if it becomes an issue.